### PR TITLE
Improve tests for `@select` and `@transform`

### PR DESCRIPTION
**Pull request by pdeffebach**
In preparation of using `DataFrames.transform` behind the scenes, I am writing a bunch of tests to 

1. Make sure we catch edge cases such as using reserved words 
2. Add `@test_throws` so we know exactly when we are adding new behavior or fixing current bugs. 

Please feel free to add tests as you see fit.

### CODE REVIEW

**Comment by pdeffebach** [comment]
Okay I think this is ready for review. 

Remaining tests are 

* `transform` with a grouped data frame. But the tests for behavior are pretty good and the macro parsing stuff should work the exact same as `@transform` with a normal data frame
* `@where` which I don't have a strategy for replacing with `filter` but whose tests could still go in this PR
* `@orderby` which is like `@where`. Tests could be added but I don't have much of an idea how behavior will change. 
* `@with`, for which more tests could be added but whose behavior is at a low risk of changing, since it is just `replace_syms!` which `@transform2` etc. use without any modification.

**Comment by bkamins** [comment]
CI fails

**Comment by pdeffebach** [comment]
Sorry, `@eval` woes. It's fixed now, feel free to have a look.

File: test/dataframes.jl [hunk 1]
```diff
@@ -4,64 +4,191 @@ module TestDataFrames
 using Test
 using DataFrames
 using DataFramesMeta
+using Statistics
+
+const ≅ = isequal
+
+df = DataFrame(
```
Comment (bkamins) [inline test/dataframes.jl:11]: it is better to wrap all variables inside a `@testset` (even if some definitions are repeated) as it makes `@testset` easier to maintain in the long run.

File: test/dataframes.jl [hunk 2]
```diff
@@ -4,64 +4,191 @@ module TestDataFrames
...
+cq = :c
+
+gr = "g"
+ir = "i"
+tr = "t"
+yr = "y"
+cr = "c"
+
+@testset "@transform" begin
+    @test @transform(df, n = :i).n == df.i
```
Comment (bkamins) [inline test/dataframes.jl:34]: in several tests I would also check:
1. if the original `df` was not changed
2. if columns got copied (`!==`) and not just aliased (`===`)
3. not only the transformed column but if all other columns are retained

**Comment by bkamins** [comment]
Nice tests - I have tried to check that you have not removed any old tests and it seems so, but just please confirm that this is the case.

File: test/dataframes.jl (hunk 2 above)
Comment (bkamins) [reply test/dataframes.jl:34]: Same for other macros

Comment (pdeffebach) [reply test/dataframes.jl:34]: > if columns got copied (`!==`) and not just aliased (`===`)

It looks like that is not actually the behavior

```
julia> df = DataFrame(rand(2,2));

julia> @transform(df, n = :x1).n === df.x1
true
```

Presumably this is not actually behavior we want, so this goes in the `limits of` testset. Do you agree?

**Comment by pdeffebach** [comment]
Okay i've pinned down some `===` behavior for `@transform` and `@select`. 

I can confirm that I didnt delete any tests. Just moved a lot of things.

File: test/dataframes.jl (hunk 2 above)
Comment (bkamins) [reply test/dataframes.jl]: Yes - for sure we want `@tranform` to make a copy like `transform`, as this is the expected behavior (so I would put the current tests in the set of tests that are likely to be broken after the update).

In general we might consider adding `copycols` kwarg to `@transform` but this can be left for later.

File: test/dataframes.jl (hunk 1 above)
Comment (pdeffebach) [reply test/dataframes.jl:65]: I've implmented that where I can but in many cases the use of `@eval` means i need the data in global scope.

Comment (bkamins) [reply test/dataframes.jl:65]: ah - right!

**Comment by pdeffebach** [comment]
Bumping this. Would be great to get this merged so we can make real changes to `@transform`

**Comment by bkamins** [comment]
Sure - but can you please resolve merge conflicts first and make sure all tests are correct against current master (without it I am not able to merge this). Thank you!

**Comment by pdeffebach** [comment]
Tests pass! Ready for merging

**Comment by bkamins** [comment]
Thank you!

**Comment by pdeffebach** [comment]
Now the real work begins!
